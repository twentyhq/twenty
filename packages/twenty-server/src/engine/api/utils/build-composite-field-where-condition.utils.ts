import { FieldMetadataType } from 'twenty-shared/types';
import { isDefined } from 'twenty-shared/utils';

import {
  ObjectRecord,
  ObjectRecordFilter,
  ObjectRecordOrderBy,
} from 'src/engine/api/graphql/workspace-query-builder/interfaces/object-record.interface';

import {
  GraphqlQueryRunnerException,
  GraphqlQueryRunnerExceptionCode,
} from 'src/engine/api/graphql/graphql-query-runner/errors/graphql-query-runner.exception';
import { computeOperator } from 'src/engine/api/utils/compute-operator.utils';
import { isAscendingOrder } from 'src/engine/api/utils/is-ascending-order.utils';
import { validateAndGetOrderByForCompositeFields } from 'src/engine/api/utils/validate-and-get-order-by.utils';
import { compositeTypeDefinitions } from 'src/engine/metadata-modules/field-metadata/composite-types';

export const buildCompositeFieldWhereCondition = ({
  fieldType,
  fieldKey,
  orderBy,
  cursorValue,
  isForwardPagination,
  operator,
}: {
  fieldType: FieldMetadataType;
  fieldKey: keyof ObjectRecord;
  orderBy: ObjectRecordOrderBy;
  cursorValue: Record<string, unknown>;
  isForwardPagination: boolean;
  operator?: string;
}): Record<string, Partial<ObjectRecordFilter>> => {
  const compositeType = compositeTypeDefinitions.get(fieldType);

  if (!compositeType) {
    throw new GraphqlQueryRunnerException(
      `Composite type definition not found for type: ${fieldType}`,
      GraphqlQueryRunnerExceptionCode.INVALID_CURSOR,
    );
  }

  const fieldOrderBy = validateAndGetOrderByForCompositeFields(
    fieldKey,
    orderBy,
  );

  const compositeFieldProperties = compositeType.properties.filter(
    (property) =>
      property.type !== FieldMetadataType.RAW_JSON &&
      cursorValue[property.name] !== undefined,
  );

  if (operator === 'eq') {
    const result: Record<string, Partial<ObjectRecordFilter>> = {};

    compositeFieldProperties.forEach((property) => {
      result[fieldKey] = {
        ...result[fieldKey],
        [property.name]: {
          eq: cursorValue[property.name],
        },
      };
    });

    return result;
  }

  if (compositeFieldProperties.length === 0) {
    return {};
  }

  if (compositeFieldProperties.length === 1) {
    const property = compositeFieldProperties[0];
    const fieldOrder = fieldOrderBy[fieldKey];

    const orderByDirection = fieldOrder?.[property.name];

    if (!isDefined(orderByDirection)) {
      throw new GraphqlQueryRunnerException(
        'Invalid cursor',
        GraphqlQueryRunnerExceptionCode.INVALID_CURSOR,
      );
    }

    const isAscending = isAscendingOrder(orderByDirection);
    const computedOperator = computeOperator(
      isAscending,
      isForwardPagination,
      operator,
    );

    return {
      [fieldKey]: {
        [property.name]: {
          [computedOperator]: cursorValue[property.name],
        },
      },
    };
  }

  const orConditions: Partial<ObjectRecordFilter>[] = [];

  for (const [index, currentProperty] of compositeFieldProperties.entries()) {
    const orderByDirection = fieldOrderBy[fieldKey]?.[currentProperty.name];

    if (!isDefined(orderByDirection)) {
      throw new GraphqlQueryRunnerException(
        'Invalid cursor',
        GraphqlQueryRunnerExceptionCode.INVALID_CURSOR,
      );
    }

    const isAscending = isAscendingOrder(orderByDirection);
    const computedOperator = computeOperator(
      isAscending,
      isForwardPagination,
      operator,
    );

    if (index === 0) {
      orConditions.push({
        [fieldKey]: {
          [currentProperty.name]: {
            [computedOperator]: cursorValue[currentProperty.name],
          },
        },
      });
    } else {
      const andConditions: Partial<ObjectRecordFilter>[] = [];

      const previousProperties = compositeFieldProperties.slice(0, index);

      for (const previousProperty of previousProperties) {
        andConditions.push({
          [fieldKey]: {
            [previousProperty.name]: {
              eq: cursorValue[previousProperty.name],
            },
          },
        });
      }

      andConditions.push({
        [fieldKey]: {
          [currentProperty.name]: {
            [computedOperator]: cursorValue[currentProperty.name],
          },
        },
      });

      orConditions.push({
        and: andConditions,
      });
    }
  }

  return {
    or: orConditions,
  };
};
