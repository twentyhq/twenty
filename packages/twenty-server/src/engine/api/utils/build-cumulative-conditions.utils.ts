import { ObjectRecordFilter } from 'src/engine/api/graphql/workspace-query-builder/interfaces/object-record.interface';

export const buildCumulativeConditions = <
  T,
  R extends ObjectRecordFilter | { and: ObjectRecordFilter[] },
>({
  items,
  buildEqualityCondition,
  buildMainCondition,
}: {
  items: T[];
  buildEqualityCondition: (item: T, index: number) => ObjectRecordFilter;
  buildMainCondition: (item: T, index: number) => ObjectRecordFilter;
}): R[] => {
  return items.map((item, index) => {
    const andConditions: ObjectRecordFilter[] = [];

    for (
      let subConditionIndex = 0;
      subConditionIndex < index;
      subConditionIndex++
    ) {
      const previousItem = items[subConditionIndex];

      andConditions.push(
        buildEqualityCondition(previousItem, subConditionIndex),
      );
    }

    andConditions.push(buildMainCondition(item, index));

    if (andConditions.length === 1) {
      return andConditions[0];
    }

    return {
      and: andConditions,
    };
  }) as R[];
};
