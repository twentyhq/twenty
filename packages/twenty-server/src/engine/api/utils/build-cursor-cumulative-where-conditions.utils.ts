import {
  ObjectRecordCursorLeafCompositeValue,
  ObjectRecordCursorLeafScalarValue,
  ObjectRecordFilter,
} from 'src/engine/api/graphql/workspace-query-builder/interfaces/object-record.interface';

type BuildCursorConditionParams<CursorValue> = {
  cursorKey: string;
  cursorValue: CursorValue;
};

type ReturnType = Array<ObjectRecordFilter | { and: ObjectRecordFilter[] }>;

type CursorEntry<CursorValue> = Record<string, CursorValue>;

type BuildCursorCumulativeWhereConditionsParams<CursorValue> = {
  cursorEntries: CursorEntry<CursorValue>[];
  buildEqualityCondition: ({
    cursorKey,
    cursorValue,
  }: BuildCursorConditionParams<CursorValue>) => ObjectRecordFilter;
  buildMainCondition: ({
    cursorKey,
    cursorValue,
  }: BuildCursorConditionParams<CursorValue>) => ObjectRecordFilter;
};

export const buildCursorCumulativeWhereCondition = <
  CursorValue extends
    | ObjectRecordCursorLeafCompositeValue
    | ObjectRecordCursorLeafScalarValue,
>({
  cursorEntries,
  buildEqualityCondition,
  buildMainCondition,
}: BuildCursorCumulativeWhereConditionsParams<CursorValue>): ReturnType => {
  return cursorEntries.map((cursorEntry, index) => {
    const [currentCursorKey, currentCursorValue] =
      Object.entries(cursorEntry)[0];
    const andConditions: ObjectRecordFilter[] = [];

    for (
      let subConditionIndex = 0;
      subConditionIndex < index;
      subConditionIndex++
    ) {
      const previousCursorEntry = cursorEntries[subConditionIndex];
      const [previousCursorKey, previousCursorValue] =
        Object.entries(previousCursorEntry)[0];

      andConditions.push(
        buildEqualityCondition({
          cursorKey: previousCursorKey,
          cursorValue: previousCursorValue,
        }),
      );
    }

    andConditions.push(
      buildMainCondition({
        cursorKey: currentCursorKey,
        cursorValue: currentCursorValue,
      }),
    );

    if (andConditions.length === 1) {
      return andConditions[0];
    }

    return {
      and: andConditions,
    };
  });
};
