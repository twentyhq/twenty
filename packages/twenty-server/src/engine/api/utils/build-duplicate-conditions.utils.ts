import isEmpty from 'lodash.isempty';

import {
  ObjectRecord,
  ObjectRecordFilter,
} from 'src/engine/api/graphql/workspace-query-builder/interfaces/object-record.interface';

import { settings } from 'src/engine/constants/settings';
import { ObjectMetadataItemWithFieldMaps } from 'src/engine/metadata-modules/types/object-metadata-item-with-field-maps';
import { getCompositeFieldMetadataMap } from 'src/engine/twenty-orm/utils/format-result.util';
import { formatData } from 'src/engine/twenty-orm/utils/format-data.util';

export const buildDuplicateConditions = (
  objectMetadataItemWithFieldMaps: ObjectMetadataItemWithFieldMaps,
  records?: Partial<ObjectRecord>[] | undefined,
  filteringByExistingRecordId?: string,
): Partial<ObjectRecordFilter> => {
  if (!records || records.length === 0) {
    return {};
  }

  const criteriaCollection =
    objectMetadataItemWithFieldMaps.duplicateCriteria || [];

  const formattedRecords = formatData(records, objectMetadataItemWithFieldMaps);

  const compositeFieldMetadataMap = getCompositeFieldMetadataMap(
    objectMetadataItemWithFieldMaps,
  );

  const conditions = formattedRecords.flatMap((record) => {
    const criteriaWithMatchingArgs = criteriaCollection.filter((criteria) =>
      criteria.every((columnName) => {
        const value = record[columnName] as string | undefined;

        return (
          value && value.length >= settings.minLengthOfStringForDuplicateCheck
        );
      }),
    );

    return criteriaWithMatchingArgs.map((criteria) => {
      const condition = {};

      criteria.forEach((columnName) => {
        const compositeFieldMetadata =
          compositeFieldMetadataMap.get(columnName);

        if (compositeFieldMetadata) {
          // @ts-expect-error legacy noImplicitAny
          condition[compositeFieldMetadata.parentField] = {
            // @ts-expect-error legacy noImplicitAny
            ...condition[compositeFieldMetadata.parentField],
            [compositeFieldMetadata.name]: { eq: record[columnName] },
          };
        } else {
          // @ts-expect-error legacy noImplicitAny
          condition[columnName] = { eq: record[columnName] };
        }
      });

      return condition;
    });
  });

  const filter: Partial<ObjectRecordFilter> = {};

  if (conditions && !isEmpty(conditions)) {
    filter.or = conditions;

    if (filteringByExistingRecordId) {
      filter.id = { neq: filteringByExistingRecordId };
    }
  }

  return filter;
};
