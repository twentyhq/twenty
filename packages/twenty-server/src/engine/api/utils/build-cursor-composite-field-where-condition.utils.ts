import { FieldMetadataType } from 'twenty-shared/types';
import { isDefined } from 'twenty-shared/utils';

import {
  ObjectRecord,
  ObjectRecordCursorLeafCompositeValue,
  ObjectRecordFilter,
  ObjectRecordOrderBy,
} from 'src/engine/api/graphql/workspace-query-builder/interfaces/object-record.interface';

import {
  GraphqlQueryRunnerException,
  GraphqlQueryRunnerExceptionCode,
} from 'src/engine/api/graphql/graphql-query-runner/errors/graphql-query-runner.exception';
import { buildCursorCumulativeWhereCondition } from 'src/engine/api/utils/build-cursor-cumulative-where-conditions.utils';
import { computeOperator } from 'src/engine/api/utils/compute-operator.utils';
import { isAscendingOrder } from 'src/engine/api/utils/is-ascending-order.utils';
import { validateAndGetOrderByForCompositeField } from 'src/engine/api/utils/validate-and-get-order-by.utils';
import { compositeTypeDefinitions } from 'src/engine/metadata-modules/field-metadata/composite-types';

type BuildCursorCompositeFieldWhereConditionParams = {
  fieldType: FieldMetadataType;
  fieldKey: keyof ObjectRecord;
  orderBy: ObjectRecordOrderBy;
  cursorValue: ObjectRecordCursorLeafCompositeValue;
  isForwardPagination: boolean;
  isEqualityCondition?: boolean;
};

export const buildCursorCompositeFieldWhereCondition = ({
  fieldType,
  fieldKey,
  orderBy,
  cursorValue,
  isForwardPagination,
  isEqualityCondition = false,
}: BuildCursorCompositeFieldWhereConditionParams): Record<
  string,
  ObjectRecordFilter
> => {
  const compositeType = compositeTypeDefinitions.get(fieldType);

  if (!compositeType) {
    throw new GraphqlQueryRunnerException(
      `Composite type definition not found for type: ${fieldType}`,
      GraphqlQueryRunnerExceptionCode.INVALID_CURSOR,
    );
  }

  const fieldOrderBy = validateAndGetOrderByForCompositeField(
    fieldKey,
    orderBy,
  );

  const compositeFieldProperties = compositeType.properties.filter(
    (property) =>
      property.type !== FieldMetadataType.RAW_JSON &&
      cursorValue[property.name] !== undefined,
  );

  if (compositeFieldProperties.length === 0) {
    return {};
  }

  const cursorEntries = compositeFieldProperties
    .map((property) => {
      if (cursorValue[property.name] === undefined) {
        return null;
      }

      return {
        [property.name]: cursorValue[property.name],
      };
    })
    .filter(isDefined);

  if (isEqualityCondition) {
    const result = cursorEntries.reduce<Record<string, ObjectRecordFilter>>(
      (acc, cursorEntry) => {
        const [cursorKey, cursorValue] = Object.entries(cursorEntry)[0];

        return {
          ...acc,
          [cursorKey]: {
            eq: cursorValue,
          },
        };
      },
      {},
    );

    return {
      [fieldKey]: result,
    };
  }

  const orConditions = buildCursorCumulativeWhereCondition({
    cursorEntries,
    buildEqualityCondition: ({ cursorKey, cursorValue }) => ({
      [fieldKey]: {
        [cursorKey]: {
          eq: cursorValue,
        },
      },
    }),
    buildMainCondition: ({ cursorKey, cursorValue }) => {
      const orderByDirection = fieldOrderBy[fieldKey]?.[cursorKey];

      if (!isDefined(orderByDirection)) {
        throw new GraphqlQueryRunnerException(
          'Invalid cursor',
          GraphqlQueryRunnerExceptionCode.INVALID_CURSOR,
        );
      }

      const isAscending = isAscendingOrder(orderByDirection);
      const computedOperator = computeOperator(
        isAscending,
        isForwardPagination,
      );

      return {
        [fieldKey]: {
          [cursorKey]: {
            [computedOperator]: cursorValue,
          },
        },
      };
    },
  });

  if (orConditions.length === 1) {
    return orConditions[0];
  }

  return {
    or: orConditions,
  };
};
