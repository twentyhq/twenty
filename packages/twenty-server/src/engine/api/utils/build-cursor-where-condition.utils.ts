import { isDefined } from 'twenty-shared/utils';

import {
  ObjectRecord,
  ObjectRecordCursorLeafCompositeValue,
  ObjectRecordCursorLeafScalarValue,
  ObjectRecordOrderBy,
} from 'src/engine/api/graphql/workspace-query-builder/interfaces/object-record.interface';

import {
  GraphqlQueryRunnerException,
  GraphqlQueryRunnerExceptionCode,
} from 'src/engine/api/graphql/graphql-query-runner/errors/graphql-query-runner.exception';
import { computeOperator } from 'src/engine/api/utils/compute-operator.utils';
import { isAscendingOrder } from 'src/engine/api/utils/is-ascending-order.utils';
import { validateAndGetOrderByForScalarField } from 'src/engine/api/utils/validate-and-get-order-by.utils';
import { isCompositeFieldMetadataType } from 'src/engine/metadata-modules/field-metadata/utils/is-composite-field-metadata-type.util';
import { FieldMetadataMap } from 'src/engine/metadata-modules/types/field-metadata-map';
import { buildCursorCompositeFieldWhereCondition } from 'src/engine/api/utils/build-cursor-composite-field-where-condition.utils';

type BuildCursorWhereConditionParams = {
  cursorKey: keyof ObjectRecord;
  cursorValue:
    | ObjectRecordCursorLeafScalarValue
    | ObjectRecordCursorLeafCompositeValue;
  fieldMetadataMapByName: FieldMetadataMap;
  orderBy: ObjectRecordOrderBy;
  isForwardPagination: boolean;
  isEqualityCondition?: boolean;
};

export const buildCursorWhereCondition = ({
  cursorKey,
  cursorValue,
  fieldMetadataMapByName,
  orderBy,
  isForwardPagination,
  isEqualityCondition = false,
}: BuildCursorWhereConditionParams): Record<string, unknown> => {
  const fieldMetadata = fieldMetadataMapByName[cursorKey];

  if (!fieldMetadata) {
    throw new GraphqlQueryRunnerException(
      `Field metadata not found for key: ${cursorKey}`,
      GraphqlQueryRunnerExceptionCode.INVALID_CURSOR,
    );
  }

  if (isCompositeFieldMetadataType(fieldMetadata.type)) {
    return buildCursorCompositeFieldWhereCondition({
      fieldType: fieldMetadata.type,
      fieldKey: cursorKey,
      orderBy,
      cursorValue: cursorValue as ObjectRecordCursorLeafCompositeValue,
      isForwardPagination,
      isEqualityCondition,
    });
  }

  if (isEqualityCondition) {
    return { [cursorKey]: { eq: cursorValue } };
  }

  const keyOrderBy = validateAndGetOrderByForScalarField(cursorKey, orderBy);
  const orderByDirection = keyOrderBy[cursorKey];

  if (!isDefined(orderByDirection)) {
    throw new GraphqlQueryRunnerException(
      'Invalid cursor',
      GraphqlQueryRunnerExceptionCode.INVALID_CURSOR,
    );
  }

  const isAscending = isAscendingOrder(orderByDirection);
  const computedOperator = computeOperator(isAscending, isForwardPagination);

  return { [cursorKey]: { [computedOperator]: cursorValue } };
};
