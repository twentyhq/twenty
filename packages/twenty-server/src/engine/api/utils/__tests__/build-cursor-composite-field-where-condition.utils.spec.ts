import { EachTestingContext } from 'twenty-shared/testing';
import { FieldMetadataType } from 'twenty-shared/types';

import {
  ObjectRecordOrderBy,
  OrderByDirection,
} from 'src/engine/api/graphql/workspace-query-builder/interfaces/object-record.interface';

import { buildCursorCompositeFieldWhereCondition } from 'src/engine/api/utils/build-cursor-composite-field-where-condition.utils';

describe('buildCompositeFieldWhereCondition', () => {
  describe('eq operator cases', () => {
    it('should handle eq operator', () => {
      const result = buildCursorCompositeFieldWhereCondition({
        fieldType: FieldMetadataType.FULL_NAME,
        fieldKey: 'name',
        orderBy: [
          {
            name: {
              firstName: OrderByDirection.AscNullsLast,
              lastName: OrderByDirection.AscNullsLast,
            },
          },
        ],
        cursorValue: { firstName: 'John', lastName: 'Doe' },
        isForwardPagination: true,
        isEqualityCondition: true,
      });

      expect(result).toEqual({
        name: {
          firstName: {
            eq: 'John',
          },
          lastName: {
            eq: 'Doe',
          },
        },
      });
    });
  });

  describe('single property cases', () => {
    const singlePropertyTestCases: EachTestingContext<{
      description: string;
      fieldType: FieldMetadataType;
      fieldKey: string;
      orderBy: ObjectRecordOrderBy;
      value: { [key: string]: any };
      isForwardPagination: boolean;
      operator?: string;
      expectedOperator: string;
    }>[] = [
      {
        title: 'ascending order with forward pagination',
        context: {
          description: 'ascending order with forward pagination',
          fieldType: FieldMetadataType.FULL_NAME,
          fieldKey: 'person',
          orderBy: [{ person: { firstName: OrderByDirection.AscNullsLast } }],
          value: { firstName: 'John' },
          isForwardPagination: true,
          operator: undefined,
          expectedOperator: 'gt',
        },
      },
      {
        title: 'ascending order with backward pagination',
        context: {
          description: 'ascending order with backward pagination',
          fieldType: FieldMetadataType.FULL_NAME,
          fieldKey: 'person',
          orderBy: [{ person: { firstName: OrderByDirection.AscNullsLast } }],
          value: { firstName: 'John' },
          isForwardPagination: false,
          operator: undefined,
          expectedOperator: 'lt',
        },
      },
      {
        title: 'descending order with forward pagination',
        context: {
          description: 'descending order with forward pagination',
          fieldType: FieldMetadataType.FULL_NAME,
          fieldKey: 'person',
          orderBy: [{ person: { firstName: OrderByDirection.DescNullsLast } }],
          value: { firstName: 'John' },
          isForwardPagination: true,
          operator: undefined,
          expectedOperator: 'lt',
        },
      },
      {
        title: 'descending order with backward pagination',
        context: {
          description: 'descending order with backward pagination',
          fieldType: FieldMetadataType.FULL_NAME,
          fieldKey: 'person',
          orderBy: [{ person: { firstName: OrderByDirection.DescNullsLast } }],
          value: { firstName: 'John' },
          isForwardPagination: false,
          operator: undefined,
          expectedOperator: 'gt',
        },
      },
    ];

    test.each(singlePropertyTestCases)(
      'should handle $description',
      ({ context }) => {
        const {
          fieldType,
          fieldKey,
          orderBy,
          value,
          isForwardPagination,
          expectedOperator,
        } = context;

        const result = buildCursorCompositeFieldWhereCondition({
          fieldType,
          fieldKey,
          orderBy,
          cursorValue: value,
          isForwardPagination,
        });

        expect(result).toEqual({
          [fieldKey]: {
            firstName: {
              [expectedOperator]: value.firstName,
            },
          },
        });
      },
    );

    test.each(singlePropertyTestCases)(
      'should match snapshot for $title',
      ({ context }) => {
        const {
          fieldType,
          fieldKey,
          orderBy,
          value,
          isForwardPagination,
          description,
        } = context;

        const result = buildCursorCompositeFieldWhereCondition({
          fieldType,
          fieldKey,
          orderBy,
          cursorValue: value,
          isForwardPagination,
        });

        expect(result).toMatchSnapshot(`single property - ${description}`);
      },
    );
  });

  describe('multiple properties cases', () => {
    const multiplePropertiesTestCases: EachTestingContext<{
      description: string;
      fieldType: FieldMetadataType;
      fieldKey: string;
      orderBy: ObjectRecordOrderBy;
      value: { [key: string]: any };
      isForwardPagination: boolean;
    }>[] = [
      {
        title: 'two properties - both ascending, forward pagination',
        context: {
          description: 'two properties - both ascending, forward pagination',
          fieldType: FieldMetadataType.FULL_NAME,
          fieldKey: 'name',
          orderBy: [
            {
              name: {
                firstName: OrderByDirection.AscNullsLast,
                lastName: OrderByDirection.AscNullsLast,
              },
            },
          ],
          value: { firstName: 'John', lastName: 'Doe' },
          isForwardPagination: true,
        },
      },
      {
        title: 'two properties - both ascending, backward pagination',
        context: {
          description: 'two properties - both ascending, backward pagination',
          fieldType: FieldMetadataType.FULL_NAME,
          fieldKey: 'name',
          orderBy: [
            {
              name: {
                firstName: OrderByDirection.AscNullsLast,
                lastName: OrderByDirection.AscNullsLast,
              },
            },
          ],
          value: { firstName: 'John', lastName: 'Doe' },
          isForwardPagination: false,
        },
      },
      {
        title: 'two properties - mixed ordering, forward pagination',
        context: {
          description: 'two properties - mixed ordering, forward pagination',
          fieldType: FieldMetadataType.FULL_NAME,
          fieldKey: 'name',
          orderBy: [
            {
              name: {
                firstName: OrderByDirection.AscNullsLast,
                lastName: OrderByDirection.DescNullsLast,
              },
            },
          ],
          value: { firstName: 'John', lastName: 'Doe' },
          isForwardPagination: true,
        },
      },
      {
        title: 'two properties - both descending, forward pagination',
        context: {
          description: 'two properties - both descending, forward pagination',
          fieldType: FieldMetadataType.FULL_NAME,
          fieldKey: 'name',
          orderBy: [
            {
              name: {
                firstName: OrderByDirection.DescNullsLast,
                lastName: OrderByDirection.DescNullsLast,
              },
            },
          ],
          value: { firstName: 'John', lastName: 'Doe' },
          isForwardPagination: true,
        },
      },
      {
        title: 'address composite field - both ascending, forward pagination',
        context: {
          description:
            'address composite field - both ascending, forward pagination',
          fieldType: FieldMetadataType.ADDRESS,
          fieldKey: 'address',
          orderBy: [
            {
              address: {
                addressStreet1: OrderByDirection.AscNullsLast,
                addressStreet2: OrderByDirection.AscNullsLast,
                addressCity: OrderByDirection.AscNullsLast,
                addressState: OrderByDirection.AscNullsLast,
                addressCountry: OrderByDirection.AscNullsLast,
                addressPostcode: OrderByDirection.AscNullsLast,
              },
            },
          ],
          value: {
            addressStreet1: '123 Main St',
            addressStreet2: 'Apt 4B',
            addressCity: 'New York',
            addressState: 'NY',
            addressCountry: 'USA',
            addressPostcode: '10001',
          },
          isForwardPagination: true,
        },
      },
    ];

    test.each(multiplePropertiesTestCases)(
      'should handle $title',
      ({ context }) => {
        const { fieldType, fieldKey, orderBy, value, isForwardPagination } =
          context;
        const result = buildCursorCompositeFieldWhereCondition({
          fieldType,
          fieldKey,
          orderBy,
          cursorValue: value,
          isForwardPagination,
        });

        expect(result).toHaveProperty('or');
        const orConditions = result.or;

        expect(Array.isArray(orConditions)).toBe(true);

        const propertiesWithValues = Object.keys(value).length;

        expect(orConditions).toHaveLength(propertiesWithValues);

        expect(orConditions[0]).toHaveProperty(fieldKey);

        for (const [index, orCondition] of orConditions.slice(1).entries()) {
          expect(orCondition).toHaveProperty('and');
          expect(Array.isArray(orCondition.and)).toBe(true);
          expect(orCondition.and).toHaveLength(index + 2);
        }
      },
    );

    test.each(multiplePropertiesTestCases)(
      'should match snapshots for $title',
      ({ context }) => {
        const {
          fieldType,
          fieldKey,
          orderBy,
          value,
          isForwardPagination,
          description,
        } = context;

        const result = buildCursorCompositeFieldWhereCondition({
          fieldType,
          fieldKey,
          orderBy,
          cursorValue: value,
          isForwardPagination,
        });

        expect(result).toMatchSnapshot(`multiple properties - ${description}`);
      },
    );
  });

  describe('error cases', () => {
    it('should throw error for invalid composite type', () => {
      expect(() =>
        buildCursorCompositeFieldWhereCondition({
          fieldType: FieldMetadataType.TEXT,
          fieldKey: 'person',
          orderBy: [{ person: { firstName: OrderByDirection.AscNullsLast } }],
          cursorValue: { firstName: 'John' },
          isForwardPagination: true,
        }),
      ).toThrow('Composite type definition not found for type: TEXT');
    });

    it('should throw error for invalid cursor with missing order by', () => {
      expect(() =>
        buildCursorCompositeFieldWhereCondition({
          fieldType: FieldMetadataType.FULL_NAME,
          fieldKey: 'person',
          orderBy: [],
          cursorValue: { firstName: 'John' },
          isForwardPagination: true,
        }),
      ).toThrow('Invalid cursor');
    });
  });
});
