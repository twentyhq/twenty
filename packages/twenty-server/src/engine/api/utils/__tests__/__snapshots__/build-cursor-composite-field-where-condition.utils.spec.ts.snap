// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildCompositeFieldWhereCondition multiple properties cases should match snapshots for address composite field - both ascending, forward pagination: multiple properties - address composite field - both ascending, forward pagination 1`] = `
{
  "or": [
    {
      "address": {
        "addressStreet1": {
          "gt": "123 Main St",
        },
      },
    },
    {
      "and": [
        {
          "address": {
            "addressStreet1": {
              "eq": "123 Main St",
            },
          },
        },
        {
          "address": {
            "addressStreet2": {
              "gt": "Apt 4B",
            },
          },
        },
      ],
    },
    {
      "and": [
        {
          "address": {
            "addressStreet1": {
              "eq": "123 Main St",
            },
          },
        },
        {
          "address": {
            "addressStreet2": {
              "eq": "Apt 4B",
            },
          },
        },
        {
          "address": {
            "addressCity": {
              "gt": "New York",
            },
          },
        },
      ],
    },
    {
      "and": [
        {
          "address": {
            "addressStreet1": {
              "eq": "123 Main St",
            },
          },
        },
        {
          "address": {
            "addressStreet2": {
              "eq": "Apt 4B",
            },
          },
        },
        {
          "address": {
            "addressCity": {
              "eq": "New York",
            },
          },
        },
        {
          "address": {
            "addressPostcode": {
              "gt": "10001",
            },
          },
        },
      ],
    },
    {
      "and": [
        {
          "address": {
            "addressStreet1": {
              "eq": "123 Main St",
            },
          },
        },
        {
          "address": {
            "addressStreet2": {
              "eq": "Apt 4B",
            },
          },
        },
        {
          "address": {
            "addressCity": {
              "eq": "New York",
            },
          },
        },
        {
          "address": {
            "addressPostcode": {
              "eq": "10001",
            },
          },
        },
        {
          "address": {
            "addressState": {
              "gt": "NY",
            },
          },
        },
      ],
    },
    {
      "and": [
        {
          "address": {
            "addressStreet1": {
              "eq": "123 Main St",
            },
          },
        },
        {
          "address": {
            "addressStreet2": {
              "eq": "Apt 4B",
            },
          },
        },
        {
          "address": {
            "addressCity": {
              "eq": "New York",
            },
          },
        },
        {
          "address": {
            "addressPostcode": {
              "eq": "10001",
            },
          },
        },
        {
          "address": {
            "addressState": {
              "eq": "NY",
            },
          },
        },
        {
          "address": {
            "addressCountry": {
              "gt": "USA",
            },
          },
        },
      ],
    },
  ],
}
`;

exports[`buildCompositeFieldWhereCondition multiple properties cases should match snapshots for two properties - both ascending, backward pagination: multiple properties - two properties - both ascending, backward pagination 1`] = `
{
  "or": [
    {
      "name": {
        "firstName": {
          "lt": "John",
        },
      },
    },
    {
      "and": [
        {
          "name": {
            "firstName": {
              "eq": "John",
            },
          },
        },
        {
          "name": {
            "lastName": {
              "lt": "Doe",
            },
          },
        },
      ],
    },
  ],
}
`;

exports[`buildCompositeFieldWhereCondition multiple properties cases should match snapshots for two properties - both ascending, forward pagination: multiple properties - two properties - both ascending, forward pagination 1`] = `
{
  "or": [
    {
      "name": {
        "firstName": {
          "gt": "John",
        },
      },
    },
    {
      "and": [
        {
          "name": {
            "firstName": {
              "eq": "John",
            },
          },
        },
        {
          "name": {
            "lastName": {
              "gt": "Doe",
            },
          },
        },
      ],
    },
  ],
}
`;

exports[`buildCompositeFieldWhereCondition multiple properties cases should match snapshots for two properties - both descending, forward pagination: multiple properties - two properties - both descending, forward pagination 1`] = `
{
  "or": [
    {
      "name": {
        "firstName": {
          "lt": "John",
        },
      },
    },
    {
      "and": [
        {
          "name": {
            "firstName": {
              "eq": "John",
            },
          },
        },
        {
          "name": {
            "lastName": {
              "lt": "Doe",
            },
          },
        },
      ],
    },
  ],
}
`;

exports[`buildCompositeFieldWhereCondition multiple properties cases should match snapshots for two properties - mixed ordering, forward pagination: multiple properties - two properties - mixed ordering, forward pagination 1`] = `
{
  "or": [
    {
      "name": {
        "firstName": {
          "gt": "John",
        },
      },
    },
    {
      "and": [
        {
          "name": {
            "firstName": {
              "eq": "John",
            },
          },
        },
        {
          "name": {
            "lastName": {
              "lt": "Doe",
            },
          },
        },
      ],
    },
  ],
}
`;

exports[`buildCompositeFieldWhereCondition single property cases should match snapshot for ascending order with backward pagination: single property - ascending order with backward pagination 1`] = `
{
  "person": {
    "firstName": {
      "lt": "John",
    },
  },
}
`;

exports[`buildCompositeFieldWhereCondition single property cases should match snapshot for ascending order with forward pagination: single property - ascending order with forward pagination 1`] = `
{
  "person": {
    "firstName": {
      "gt": "John",
    },
  },
}
`;

exports[`buildCompositeFieldWhereCondition single property cases should match snapshot for descending order with backward pagination: single property - descending order with backward pagination 1`] = `
{
  "person": {
    "firstName": {
      "gt": "John",
    },
  },
}
`;

exports[`buildCompositeFieldWhereCondition single property cases should match snapshot for descending order with forward pagination: single property - descending order with forward pagination 1`] = `
{
  "person": {
    "firstName": {
      "lt": "John",
    },
  },
}
`;
